linode/manager
# upcoming: [DI-21200] - Migration from ChartJS to Recharts in CloudPulse

## Description 📝
Migrated ChartJS to Recharts in CloudPulse.

Note: This PR is raised so that code changes can be reviewed & parallelly we are working on automation test cases which will eventually be added in this PR.

## Changes  🔄
List any change relevant to the reviewer.
1. Replaced LineGraph with AreaChart in CloudPulseLineGraph component.
2. Updated x-axis ticks & chart variant in CloudPulseWidget
3. Modified structure the of data passed to AreaChart component in CloudPulseWidgetUtils
4. Removed CloudPulseColorPalette
5. Modified legendColor property in metrics display to accept any color
6. Added a connectNulls property in AreaChart component
7. Added DataSet type in cloudpulse types

## Target release date 🗓️
18th November

## Preview 📷
**Include a screenshot or screen recording of the change.**

:lock: Use the [Mask Sensitive Data](https://cloud.linode.com/profile/settings) setting for security.

:bulb: Use `<video src="" />` tag when including recordings in table.

No Data Display
![Screenshot 2024-11-04 at 1 41 57 PM](https://github.com/user-attachments/assets/a18d0dd6-025d-41bd-b386-2f1785a6fcca)
![Screenshot 2024-11-04 at 1 42 10 PM](https://github.com/user-attachments/assets/63794c06-9aa7-4539-a356-035806e9b688)

X-Axis with timestamp ticks for less than 24hrs
![Screenshot 2024-11-04 at 1 42 22 PM](https://github.com/user-attachments/assets/7fa7b0aa-a9e4-4aff-92cc-5c7dd8d5162a)
![Screenshot 2024-11-04 at 1 42 30 PM](https://github.com/user-attachments/assets/6731f28a-2a6b-4c7d-b7b1-944c45dfb0f9)

X-Axis with date & timestamp ticks for more than 24hrs
![Screenshot 2024-11-04 at 1 42 57 PM](https://github.com/user-attachments/assets/4e03c854-9ff0-4599-969c-551516ca9761)
![Screenshot 2024-11-04 at 1 43 16 PM](https://github.com/user-attachments/assets/5515a9bd-2b2e-401b-8dc1-0576d4a6aa45)

Legends rows 
https://github.com/user-attachments/assets/bc8886ba-db7a-44ba-aa09-fea7d715bd22

Error State
![Screenshot 2024-11-04 at 2 16 32 PM](https://github.com/user-attachments/assets/c006e94c-055b-4537-b55b-ca5dfe5fd70a)
![Screenshot 2024-11-04 at 2 16 40 PM](https://github.com/user-attachments/assets/122830ba-9917-4e45-aad9-b0e965eb76fd)


## How to test 🧪

1. Switch to mock user
2. Select mandatory filters to generate the graphs
3. Change time duration filter to see difference in x-axis tick
4. Modify data in serverHandlers.ts for `monitor/services/:serviceType/metrics` endpoint to playaround with graph & legends.
5. You can see a scroll bar in legend rows table if more than 2 legends are present



## As an Author I have considered 🤔

*Check all that apply*
